### PR TITLE
First hack at directly accessing the item network.

### PR DESCRIPTION
**Pull request by bengardner**
If the cursor is empty:
left-click takes one item.
shift+left-click takes on stack.
control+left-click takes everything.

If the cursor is not empty:
left-click dumps the cursor content into the network.

### CODE REVIEW

**Comment by bengardner** [comment]
Fixed #54

File: src/NetworkViewUi.lua [hunk 1]
```diff
@@ -163,45 +181,125 @@ function M.update_items(player_index)
...
 
-  local rows = M.get_rows_of_items(net_view.view_type)
+  local h_stack_def = {
+    type = "flow",
+    direction = "horizontal",
+  }
+
+  -- pad the item row so there is a slot to dump items from the cursor
+  local did_pad = false
+  local function pad_item_row(item_h_stack, count)
```
Comment (year6b7a) [inline src/NetworkViewUi.lua]: What do you think about adding the pad item sprite as the first cell in the grid rather than the last row?

File: src/NetworkViewUi.lua [hunk 2]
```diff
@@ -163,45 +181,125 @@ function M.update_items(player_index)
...
+    if is_item then
+      for _ = 1, count do
+        item_h_stack.add({
+          type = "sprite-button",
+          sprite = "utility/slot_icon_resource_black",
+          tags = { event = UiConstants.NV_ITEM_SPRITE },
```
Comment (year6b7a) [inline src/NetworkViewUi.lua:225]: Can this have a tooltip to explain what it is used for?

File: src/NetworkViewUi.lua [hunk 3]
```diff
@@ -136,6 +136,21 @@ local tab_idx_to_view_type = {
   "shortage",
 }
 
+local function item_tooltip(name, count)
+  return { "", game.item_prototypes[name].localised_name, ": ", count }
```
Comment (year6b7a) [inline src/NetworkViewUi.lua]: What do you think about adding instructions for accessing items to the tooltip?

File: src/NetworkViewUi.lua [hunk 4]
```diff
@@ -163,45 +181,125 @@ function M.update_items(player_index)
...
+
+  else
+    -- There is a stack in the cursor. Deposit it.
+    local cs = player.cursor_stack
+    if not cs or not cs.valid_for_read then
+      return
+    end
+
+    -- don't deposit tracked entities (can be unique)
+    if cs.item_number ~= nil then
```
Comment (year6b7a) [inline src/NetworkViewUi.lua:311]: Is this check for unique entities like spidertrons or armor? Nice catch. Currently network chests don't check for this so I'm thinking if this should be added there.

Comment (bengardner) [reply src/NetworkViewUi.lua:311]: Yeah, I was thinking what would happen if I put my decked-out Spidertron in the item-network. All gone.
I don't know how to fix that, short of implementing a true inventory, so I just added protection here.
I'd be OK with preventing the addition of anything in a non-clean state. I just didn't know how.  I suppose we could check for a custom name and a non-empty inventory (grid/trunk/logistic/modules/ammo/etc).
We could also dump all inventories when adding a complex item to the item network.

File: src/NetworkViewUi.lua (hunk 3 above)
Comment (bengardner) [reply src/NetworkViewUi.lua]: "left click with an empty cursor to remove items"?

File: src/NetworkViewUi.lua (hunk 2 above)
Comment (bengardner) [reply src/NetworkViewUi.lua:225]: Yes. What sort of wording?
Perhaps something like "click to crop the cursor inventory into the item network"?
Might be too wordy.

File: src/NetworkViewUi.lua (hunk 1 above)
Comment (bengardner) [reply src/NetworkViewUi.lua]: I'd like that.  I skipped that because I'd have to add it to the "get list" function. Might not be a bad idea.
I also tried clicking anywhere in the background, but I couldn't get that to work.

File: src/NetworkViewUi.lua (hunk 2 above)
Comment (year6b7a) [reply src/NetworkViewUi.lua:225]: That sounds like a great start. If you get something in, I'll get it merged and adjust the language if needed.

File: src/NetworkViewUi.lua (hunk 3 above)
Comment (year6b7a) [reply src/NetworkViewUi.lua]: I like the way that Factory Planner labels each key command on a separate line. I'd prefer to list every available key combination rather than try to keep the description short.

image

Let me know how much of this you want to implement before I merge this PR and finish it up.

File: src/NetworkViewUi.lua (hunk 4 above)
Comment (year6b7a) [reply src/NetworkViewUi.lua:311]: For now I'd say it's not a huge priority. I haven't had a single bug ticket around this behavior so I'm fine leaving it as-is and waiting for someone to report it as an issue.

Worst case scenario, people can always just re-load the last autosave.

File: src/NetworkViewUi.lua (hunk 3 above)
Comment (bengardner) [reply src/NetworkViewUi.lua]: Feel free to update the tooltip text.  I'm likely not going to have time to play with this again until the weekend.

**Comment by bengardner** [comment]
Rebased.

**Comment by year6b7a** [comment]
Merged! All I changed was I cleaned up the tooltips, refactored the sprite generation logic and added the "deposit into network" button in. I also fixed a bug where the view couldn't handle invalid item names.
